Add custom panel for KNX integration

marvin-w opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
* Adds the custom panel to the KNX core integration
* Adds websocket commands for the KNX custom panel to the KNX integration.

Open tasks:

- [ ] Hash the resulting frontend artefact so it wont get cached
- [ ] Update panel to use correct websocket commands (knx/ and not knx_panel/)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

marvin-w committed:
Adds websocket commands for KNX integration

marvin-w committed:
Test invalid payload type for websocket subscription

balloob commented:
Would it make sense to wait getting this merged until the custom panel PyPI package is also added to the manifest ?

marvin-w commented:
> Would it make sense to wait getting this merged until the custom panel PyPI package is also added to the manifest ?

We can do that, but it doesn't really hurt anyone if we merge it (and the diff would be smaller probably).

balloob commented:
The reason I mention it is that while working on the panel, you'll probably realize you need more or slightly different APIs. If you're fine with the current code, hit merge!

marvin-w committed:
Add custom panel to KNX integration

marvin-w commented:
> 

Makes total sense! I converted this into a draft for now and added the first running version of the panel here as well. Once this is somewhat usable I'll open it again. 

As of the APIs, we kind of intended to improve this overtime, so the panel of course be usable when I reopen this but there are still some features that we want to add later on.

As for the planned features:

- Get information about the KNX bus state, running, not running, currently assigned Individual Address etc.
- Bus Monitor where users see the Telegram flow (like objects that change the state of their smart home devices)
- Upload and parse KNX project files
- Show project information in the UI
- Support defining entities in the UI (in addition to YAML)

Currently, when a user wants to use KNX he/she mostly needs to buy a proprietary software to open the project files (it's 1000 EUR per license). Another feature we intend to add later on is the ability to upload the so called ".knxproj" file which basically is a ZIP file containing XML that we can parse with https://github.com/XKNX/xknxproject (its a start, not yet fully done) and then maybe export some JSON containing all the information we need to show them nicely in the UI.

And, as an ultimate goal it would be awesome if we can even add entities through that panel since the user experience would be so much better with the information from the KNX Project files available (all those "1/3/4" and "1/4/5" have different types, like bool, string, percent etc that we can read from the project and then when defining the entities only show those that are relevant to the user and he/she can just select it). We just need to find a way on how to do this via a websocket command or similar and store the entity information somewhere.

Obviously, not all of that can be done in a months time so we plan to ship it with the first two features and then extend it over time.

marvin-w commented on homeassistant/components/knx/__init__.py:
```diff
@@ -214,6 +216,27 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    hass.components.frontend.async_register_built_in_panel(
+        component_name="custom",
+        sidebar_title="KNX UI",
+        sidebar_icon="mdi:earth",
```
Change this to something that makes "more sense" for KNX

balloob commented on homeassistant/components/knx/__init__.py:
```diff
@@ -214,6 +216,27 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+
+    register_websocket_api(hass)
+
+    url = "/api/panel_custom/knx_ui"
```
For Insteon we did `static_insteon`. Let's stay consistent.

balloob commented on homeassistant/components/knx/__init__.py:
```diff
@@ -214,6 +216,27 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    hass.http.register_static_path(url, get_knx_ui())
+    hass.components.frontend.async_register_built_in_panel(
+        component_name="custom",
+        sidebar_title="KNX UI",
```
```suggestion
        sidebar_title="KNX",
```

balloob commented on homeassistant/components/knx/websocket.py:
```diff
@@ -0,0 +1,106 @@
+        vol.Required("type"): "knx/info",
+    }
+)
+@websocket_api.async_response
+async def ws_info(
```
Doesn't need to be async.
```suggestion
@callback
def ws_info(
```

balloob commented on homeassistant/components/knx/websocket.py:
```diff
@@ -0,0 +1,106 @@
+    }
+)
+@websocket_api.async_response
+async def ws_subscribe_telegram(
```
Doesn't need to be async.

marvin-w committed:
Apply review suggestions

marvin-w committed:
Correct dependencies in manifest

balloob commented on homeassistant/components/knx/websocket.py:
```diff
@@ -0,0 +1,135 @@
+            webcomponent_name="knx-panel",
+            sidebar_title=DOMAIN.upper(),
+            sidebar_icon="mdi:earth",
+            module_url=f"{URL_BASE}",
```
This should point at the JavaScript, but it now points ` /static_knx` ? It should be the same URL as registered for trhe static path?

marvin-w replied:
Yep, still need to fix this. Currently I just map `/static_knx` to the final JS: https://github.com/XKNX/custom-panel/blob/main/xknx_custom_panel/__init__.py#L7

Which is a bit weird - so will change that. I also have to add the hashing still.

balloob reviewed (approved):
Looks good. Please add a test that you can fetch the JS from the custom component

marvin-w commented:
> Looks good. Please add a test that you can fetch the JS from the custom component

Will do! This PR is not yet fully done. The old panel was relying on existing web components like ha-data-table which are not accessible in an iframe - so will need some more work to get back to that feature level.

marvin-w commented:
I'm on it
